### PR TITLE
api: add impl_terminator annotation and leverage in interpretation

### PR DESCRIPTION
**Pull request by superlopuh**

### CODE REVIEW

File: docs/Toy/toy/interpreter.py [hunk 1]
```diff
@@ -98,11 +107,12 @@ def run_mul(
 
         return (Tensor([l * r for l, r in zip(lhs.data, rhs.data)], lhs.shape),)
 
-    @impl(toy.ReturnOp)
+    @impl_terminator(toy.ReturnOp)
     def run_return(
         self, interpreter: Interpreter, op: toy.ReturnOp, args: tuple[Any, ...]
-    ) -> tuple[Any, ...]:
-        return args
+    ) -> tuple[TerminatorValue, PythonValues]:
```
Comment (math-fehr) [inline docs/Toy/toy/interpreter.py:113]: What is the use of `PythonValues` here?
Is it in the case where the terminator has return values (which is cursed, but allowed)?

File: xdsl/interpreter.py [hunk 2]
```diff
@@ -331,8 +388,43 @@ def interpreter_assert(self, condition: bool, message: str | None = None):
...
+
+
+class ReturnedValues(NamedTuple):
+    values: PythonValues
+
+
+TerminatorValue: TypeAlias = ReturnedValues
+
+
+class OpImplResult:
```
Comment (math-fehr) [inline xdsl/interpreter.py:401]: Can you document this, and make it a dataclass?

**Review by math-fehr — APPROVED** [review]
Nice :+1:

File: docs/Toy/toy/interpreter.py (hunk 1 above)
Comment (superlopuh) [reply docs/Toy/toy/interpreter.py:113]: Exactly, for spooky uses that I can't think of, but technically could happen

File: xdsl/interpreter.py (hunk 2 above)
Comment (superlopuh) [reply xdsl/interpreter.py]: made it a NamedTuple instead

Comment (math-fehr) [reply xdsl/interpreter.py]: Makes sense!

**Review by math-fehr — APPROVED** [review]
Thanks!

**Review by PapyChacal — APPROVED** [review]

File: xdsl/interpreter.py [hunk 3]
```diff
@@ -81,9 +91,52 @@ def impl(
...
+        ) -> OpImplResult:
+            return OpImplResult(func(ft, interpreter, op, values), None)
+
+        setattr(impl, _IMPL_OP_TYPE, op_type)
+        return impl
+
+    return annot
+
+
+def impl_terminator(
```
Comment (compor) [inline xdsl/interpreter.py:113]: 👍🏽

**Pull request merged**